File: app/src/main/java/com/maxandron/daggerissue325java/FooProvider.java
package com.maxandron.daggerissue325java;

import dagger.Module;
import dagger.Provides;

@Module
public class FooProvider {
    private final FooActivity view;

    public FooProvider(FooActivity view) {
        this.view = view;
    }

    @Provides
    @FooScope
    FooActivity provideView() {
        return view;
    }

    @Provides
    @FooScope
    Foo provideFooOrBar() {
        return new Foo();
    }
}

File: app/src/main/java/com/maxandron/daggerissue325java/foo/FooProvider.kt
package com.maxandron.daggerissue325java.foo

import dagger.Module
import dagger.Provides

@Module
class FooProvider(private val view: FooActivity) {

    @Provides @FooScope fun provideView() = view

    @Provides @FooScope fun provideFooOrBar() = Foo()
}
